src/Bicep.LangServer.IntegrationTests/Helpers/ServerRequestHelper.cs

```diff
@@ -97,6 +98,50 @@ public BicepFile ApplyCompletion(CompletionItem completion, params string[] tabS
 
             return SourceFileFactory.CreateBicepFile(bicepFile.FileUri, replaced);
         }
+
+        public async Task<LocationLink> GotoDefinition(int cursor)
+        {
+            var response = await client.RequestDefinition(new DefinitionParams
+            {
+                TextDocument = new TextDocumentIdentifier(bicepFile.FileUri),
+                Position = TextCoordinateConverter.GetPosition(bicepFile.LineStarts, cursor)
+            });
+
+            response.Should().NotBeNull();
+
+
+            // go to def should produce single result in all cases
+            response.Should().HaveCount(1);
+            var single = response.Single();
+
+            single.IsLocation.Should().BeFalse();
+            single.IsLocationLink.Should().BeTrue();
+
+            single.Location.Should().BeNull();
+            single.LocationLink.Should().NotBeNull();
+
+            return single.LocationLink!;
```

Comment:
`LocationLink` is a nullable property - the `!` converts the type from `<type> | null` to `<type>`.

Generally C# is smart about type analysis - for example in the following case, the `!` would be unneccessary:
```csharp
if (single.LocationLink is null)
{
    throw new ArgumentNullException();
}

// the C# compiler knows from the above "if" block that there's no way LocationLink can be null if execution reached this point
return single.LocationLink;
```

However, `single.LocationLink.Should().NotBeNull();` is a function supplied by an external library (`FluentAssertions`). As far as the C# compiler is concerned, this is just a function call - it doesn't know that it's going to have a similar impact on code execution as my example above.

I would also say as a general rule that `!` in **product code** is usually best avoided as it can lead to some unsafe scenarios. In test code, however, it's often unavoidable.

Reply:
Makes sense!

Reply:
Btw - IIRC we can just do

```C#
return single.LocationLink.Should().NotBeNull().Subject;
```

Reply:
An interesting option here: https://github.com/fluentassertions/fluentassertions/issues/1115#issuecomment-925869230

Reply:
Ha, that's an interesting one.